pygatt/bled112_backend.py

```diff
@@ -0,0 +1,1977 @@
+)
+from constants import LOG_FORMAT, LOG_LEVEL
+
+
+class Characteristic(object):
+    """
+    GATT characteristic. For internal use within BLED112Backend.
+    """
+    def __init__(self, name, handle):
+        """
+        Sets the characteritic name and handle.
+        """
+        self.handle = handle  # bytearray
+        self.descriptors = {
+            # uuid_string: handle
+        }
+
+    def add_descriptor(self, uuid, handle):
+        """
+        Add a characteristic descriptor to the dictionary of descriptors.
+        """
+        self.descriptors[uuid] = handle
+
+
+class AdvertisingAndScanInfo(object):
+    """
+    Holds the advertising and scan response packet data from a device at a given
+    address.
+    """
+    def __init__(self):
+        self.name = ""
+        self.address = ""
+        self.packet_data = {
+            # scan_response_packet_type[xxx]: data_dictionary,
+        }
+
+
+class BLED112Backend(object):
+    """
+    Pygatt BLE device backend using the Bluegiga BLED112.
+    Only supports 1 device connection at a time.
+    This object is NOT threadsafe.
+    """
+    def __init__(self, serial_port, run=True, logfile=None):
+        """
+        Initialize the BLED112 to be ready for use with a BLE device, i.e.,
+        stop ongoing procedures, disconnect any connections, optionally start
+        the receiver thread, and optionally delete any stored bonds.
+
+        serial_port -- The name of the serial port that the BLED112 is connected
+                       to.
+        run -- begin reveiving packets immediately.
+        logfile -- the file to log to.
+
+        Locking total order:
+        1) self._cond
+        2) self._loglock
+        """
+        # Set up logging
+        self._loglock = threading.Lock()
+        self._logger = logging.getLogger(__name__)
+        self._logger.setLevel(LOG_LEVEL)
+        if logfile is not None:
+            handler = logging.FileHandler(logfile)
+        else:  # print to stderr
+            handler = logging.StreamHandler()
+        formatter = logging.Formatter(fmt=LOG_FORMAT)
+        handler.setLevel(LOG_LEVEL)
+        handler.setFormatter(formatter)
+        self._logger.addHandler(handler)
+
+        # Initialization
+        self._lib = bled112_bglib.BGLib(loghandler=handler,
+                                        loglevel=LOG_LEVEL)
+        # Note: _ser is not protected by _main_thread_cond
+        self._ser = serial.Serial(serial_port, timeout=0.25)
+
+        # Main thread (the one calling commands) waits on this. This also
+        # provides mutual exclustion for BLED112Backend's state (except for the
+        # log)
+        self._main_thread_cond = threading.Condition()
+
+        # Packet receiving
+        self._recvr_thread = None  # background thread to receive packets
+        self._recvr_stop = True  # tell recvr thread not to run
+
+        # Flags to tell the main thread what to do when woken up from waiting
```

Comment:
This is a lot of internal state.  Can any of it be removed and encapsulated elsewhere?